docs/topics/guideline/solidity.md

```diff
@@ -0,0 +1,307 @@
+
+### TO DISCUSS: Use of Libraries (except for the all-internal libraries)
+
+## Upgradeable contracts
+
+### Leave some space in the storage for future variables at the end
+
+```
+// Reserved storage space to allow for layout changes in the future.
+uint256[50] private ______gap;
+```
+
+### Use a separate contract for storage
+
+Create one storage contract and then inherit it into the implementation. This way it is easier to manage the storage and
+add variables during upgrade.
+
+### Loop length must be short and controlled
+
+All the iterations in the smart contract must be limited in their length. A permanent denial of service could happen if
+the gas consumed by a loop is too high.
+
+### Rounding
+
+Each time a division is done the numbers are rounded (the operation is like a `floor(a/b)`) and there is some remainder
+to take into account.
+
+## Visibility
+
+### TO DISCUSS: Use of private attributes
+
+Private attributes just hide the internal state of the contract, child contracts (via inheritance), and make writing
+tests impossible.
+
+***When possible use internal or public.***
+
+Internal solve the testing issue. A mock child contract can be used to access the variable.
+
+Public solve all the issues the only downside is that it adds a default getter and calling a public method can make the
+contract code and gas consumption higher.
+
+### TO DISCUSS: Use of private view methods
+
+Usually view methods are used as a getters or to do some reusable calculation. Exposing the method helps to test the
+contract and is useful to check the internal state of a deployed contract. Methods can be exposed using public or
+external. If possible prefer external because external function's parameters are not copied into memory but are read
+directly from `calldata` and this can save some gas.
+
+### Calling internal functions is cheaper
+
+Calling internal functions is cheaper than calling public functions. When you call a public function all the parameters
+are copied again into memory and passed to that function. By contrast, when you call an internal function, references of
+those parameters are passed, and they are not copied again.
+
+## Others
+
+### Lock pragmas to specific compiler version
+
+Pragma must select a specific version of the compiler, instead of using `pragma solidity ^0.8.2;`
+use `pragma solidity 0.8.2;`.
+
+### Use interface type instead of the address for type safety
+
+```
+interface SomethingToCall {
+    function call() returns (bool);
+}
+
+contract Example {
+    function Good(SomethingToCall _something) internal returns (bool) {
+    }
+
+    function Bad(address _addr) internal returns (bool) {
+        SomethingToCall something = SomethingToCall(_addr);
+    }
+}
+```
+
+### Use short reason strings
+
+The solidity linter warns about it. Reason strings (those used in `revert`) that takes more than 32 bytes make the
+contract byte code bigger.
+
+### TO DISCUSS: Pull over Push
```

Comment:
It could be great if our experienced Solidity dev could lead us on this pattern

Reply:
no idea